banyan/src/forest/index_iter.rs

```diff
@@ -0,0 +1,208 @@
+                        };
+                        self.query
+                            .intersecting(start_offset, index, &mut head.filter);
+                        debug_assert_eq!(branch.children.len(), head.filter.len());
+
+                        break head.index.clone();
+                    }
+
+                    let next_idx = head.position as usize;
+                    if head.filter[next_idx] {
+                        // Descend into next child
+                        self.stack.push(TraverseState::new(Arc::new(
+                            branch.children[next_idx].clone(),
+                        )));
+                        continue;
+                    } else {
+                        let index = &branch.children[next_idx];
+                        match self.mode {
+                            Mode::Forward => {
+                                self.offset += index.count();
+                            }
+                            Mode::Backward => {
+                                self.offset -= index.count();
+                            }
+                        }
+                        head.next_pos(&self.mode);
+                    }
+                }
+
+                Ok(NodeInfo::Leaf(index, _)) => {
+                    match self.mode {
+                        Mode::Forward => {
+                            self.offset += index.keys.count();
+                        }
+                        Mode::Backward => {
+                            self.offset -= index.keys.count();
+                        }
+                    }
+                    // Ascend to parent's node, if it exists
+                    let this_index = self.stack.pop().expect("not empty").index;
+                    if let Some(last) = self.stack.last_mut() {
+                        last.next_pos(&self.mode);
+                    }
+                    break this_index;
+                }
+
+                // even for purged leafs and branches or ignored chunks,
+                // produce a placeholder.
+                Ok(_) => {
+                    let TraverseState { index, .. } = self.stack.pop().expect("not empty");
+                    // Ascend to parent's node. This might be none in case the
+                    // tree's root node is a `PurgedBranch`.
+                    if let Some(last) = self.stack.last_mut() {
+                        last.next_pos(&self.mode);
+                    };
+                    match self.mode {
+                        Mode::Forward => {
+                            self.offset += index.count();
+                        }
+                        Mode::Backward => {
+                            self.offset -= index.count();
+                        }
+                    };
+                    break index;
+                }
+                Err(e) => return Some(Err(e)),
+            };
+        };
+        Some(Ok(res))
+    }
+}
```

Comment:
That's basically a stripped down version of `ForestIter`. I refrained from DRYing it. Maybe I will do it to refactor retain, but for now I want to use that for weight based pruning of ephemeral events.

Reply:
Well, if it was OK for the other PR it will be OK for this. Although it sure would be nice if we could get rust to build the state machine for us...